KAFKA-9579 Fetch implementation for records in the remote storage through a specific purgatory.

satishd opened the pull request:
This PR includes 
- Recognize the fetch requests with out of range local log offsets
- Add fetch implementation for the data lying in the range of [logStartOffset, localLogStartOffset]
- Add a new purgatory for async remote read requests which are served through a specific thread pool

We have an extended version of remote fetch that can fetch from multiple remote partitions in parallel, which we will raise as a followup PR.

A few tests for the newly introduced changes are added in this PR. There are some tests available for these scenarios in 2.8.x, refactoring with the trunk changes, will add them in followup PRs.

Other contributors: 
kamal.chandraprakash@gmail.com - Further improvements and adding a few tests
showuon@gmail.com - Added a few test cases for these changes.

PS: This functionality is pulled out from internal branches with other functionalities related to the feature in 2.8.x. The reason for not pulling all the changes as it makes the PR huge, and burdensome to review and it also needs other metrics, minor enhancements(including perf), and minor changes done for tests. So, we will try to have followup PRs to cover all those.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

satishd committed:
Initial versionof remote consumer fetch implementation.

satishd committed:
Added other remote read changes in RLM and its dependant classes through a purgatory

satishd committed:
More cleanup on remote fetch path

showuon committed:
Add default value as None for delayedRemoteFetchPurgatory in the arguments and init it.

showuon committed:
add remoteLogReader tests

satishd committed:
Fixed build failures after pulling the trunk changes

satishd committed:
Minor code cleanup including simpifying the code

showuon commented on core/src/main/java/kafka/log/remote/RemoteLogManager.java:
```diff
@@ -619,6 +811,49 @@ long findHighestRemoteOffset(TopicIdPartition topicIdPartition) throws RemoteSto
+     * A remote log read task returned by asyncRead(). The caller of asyncRead() can use this object to cancel a
+     * pending task or check if the task is done.
+     */
+    public class AsyncReadTask {
```
I don't see `AsyncReadTask` is used anywhere in the code. Could we remove it?

showuon commented on core/src/main/java/kafka/log/remote/RemoteLogManager.java:
```diff
@@ -600,6 +622,176 @@ public String toString() {
+        if (!rlsMetadata.isPresent()) {
+            String epochStr = (epoch.isPresent()) ? Integer.toString(epoch.getAsInt()) : "NOT AVAILABLE";
+            throw new OffsetOutOfRangeException("Received request for offset " + offset + " for leader epoch "
+                    + epochStr + " and partition " + tp + " which does not exist in remote tier. Try again later.");
```
Are we sure `Try again later` can fix this issue? Especially, it's from application, this log doesn't help at all. Could we remove it? Thoughts?

showuon commented on core/src/main/java/kafka/log/remote/RemoteLogReader.java:
```diff
@@ -0,0 +1,70 @@
+    public Void call() {
+        RemoteLogReadResult result;
+        try {
+            logger.debug("Reading remote bytes for {}", fetchInfo.topicPartition);
```
The `remote bytes` is confusing me. How about `Reading [records] from remote storage for {}`?

showuon commented on core/src/main/java/kafka/log/remote/RemoteLogReader.java:
```diff
@@ -0,0 +1,70 @@
+            result = new RemoteLogReadResult(Optional.empty(), Optional.of(e));
+        }
+
+        logger.debug("Finished reading remote bytes for {}", fetchInfo.topicPartition);
```
ditto

showuon commented on core/src/main/scala/kafka/server/DelayedRemoteFetch.scala:
```diff
@@ -0,0 +1,121 @@
+            return forceComplete()
+        }
+    }
+    if (remoteFetchResult.isDone)
```
should we add `case C` for this case?

showuon commented on core/src/main/scala/kafka/server/ReplicaManager.scala:
```diff
@@ -1083,48 +1095,100 @@ class ReplicaManager(val config: KafkaConfig,
+            return
+        }
+
+        // If there is remote data, we will read remote data, instead of waiting for new data.
```
I think this comment should be put on L1100 after `if (remoteFetchInfo.isPresent) {`

satishd replied:
No, this comment is for the next 2 lines, mainly for the below line. 

```
delayedRemoteFetchPurgatory.tryCompleteElseWatch(remoteFetch, Seq(key))
```

showuon commented on core/src/main/scala/kafka/server/ReplicaManager.scala:
```diff
@@ -1196,17 +1252,45 @@ class ReplicaManager(val config: KafkaConfig,
+          // In case of offset out of range errors, check for remote log manager for non-compacted topics
+          // to fetch from remote storage. `log` instance should not be null here as that would have caught earlier with
+          // NotLeaderForPartitionException or ReplicaNotAvailableException.
+          // If it is from a follower then send the offset metadata but not the records data as that can be fetched
```
nit: `as [the follower] can be fetched from the remote store directly`

satishd replied:
I reworded that sentence to make it more clear.

showuon reviewed (commented):
Thanks for the PR. Left some comments.

satishd committed:
Addressed review comments

junrao commented on core/src/main/scala/kafka/server/ReplicaManager.scala:
```diff
@@ -1083,48 +1095,100 @@ class ReplicaManager(val config: KafkaConfig,
-      // may arrive and hence make this operation completable.
-      delayedFetchPurgatory.tryCompleteElseWatch(delayedFetch, delayedFetchKeys)
+
+      if (remoteFetchInfo.isPresent) {
```
In line 1082, we should further test `!remoteFetchInfo.isPresent`, right?

satishd replied:
I am not sure line num:1082 is sane as you meant it to be as the file could have been updated. Please clarify.

junrao replied:
In the following code, we should go into that branch only if `remoteFetchInfo` is empty, right? Otherwise, if we could get into a situation that a remote partition is never served because the fetch request  is always satisfied with new local data on other partitions.
```
    if (params.maxWaitMs <= 0 || fetchInfos.isEmpty || bytesReadable >= params.minBytes || errorReadingData ||
      hasDivergingEpoch || hasPreferredReadReplica) {
```

satishd replied:
Do you mean to say that we should not return immediately if `remoteFetchInfo` exists because that should be served otherwise remote fetches may starve as long as there is enough data immediately available to be sent? So, the condition becomes 

```
    if (!remoteFetchInfo.isPresent && (params.maxWaitMs <= 0 || fetchInfos.isEmpty 
            || bytesReadable >= params.minBytes || errorReadingData || hasDivergingEpoch 
            || hasPreferredReadReplica))
```

junrao replied:
Yes

satishd replied:
Sure, that check was missed while pulling the changes. Good catch. Updated it with the latest commit.

junrao commented on core/src/main/scala/kafka/server/ReplicaManager.scala:
```diff
@@ -1243,6 +1327,33 @@ class ReplicaManager(val config: KafkaConfig,
     result
   }
 
+  def createLogReadResult(highWatermark: Long,
```
Could this be private? Ditto below.

satishd replied:
`createLogReadResult(e: Throwable)` can not be private as it is used in `DelayedRemoteFetch`. But this method can be used. It is going to be used in test classes that we are going to add in this PR or followup PR.

junrao commented on core/src/main/scala/kafka/server/DelayedRemoteFetch.scala:
```diff
@@ -0,0 +1,121 @@
+                         fetchParams: FetchParams,
+                         localReadResults: Seq[(TopicIdPartition, LogReadResult)],
+                         replicaManager: ReplicaManager,
+                         quota: ReplicaQuota,
```
quota seems unused.

showuon commented on core/src/main/scala/kafka/server/ReplicaManager.scala:
```diff
@@ -1061,6 +1072,7 @@ class ReplicaManager(val config: KafkaConfig,
       logReadResultMap.put(topicIdPartition, logReadResult)
     }
 
+    val delayedRemoteFetchPurgatory: DelayedOperationPurgatory[DelayedRemoteFetch] = null;
```
Why should we make `delayedRemoteFetchPurgatory` null here? We've init `delayedRemoteFetchPurgatory` in L217, and want to use it below L1131 immediately. So this line should be removed.

showuon replied:
OK, I've run a test, and confirmed this line will make it throw NPE while fetching data from remote segment.

showuon commented on core/src/main/java/kafka/log/remote/RemoteLogManager.java:
```diff
@@ -600,6 +622,176 @@ public String toString() {
+            RecordBatch firstBatch = findFirstBatch(remoteLogInputStream, offset);
+
+            if (firstBatch == null)
+                return new FetchDataInfo(new LogOffsetMetadata(offset), MemoryRecords.EMPTY, false,
```
I think we need to log something in this case.

satishd committed:
Resolved conflicts ancountered from pulling the trunk

jeqo commented on core/src/main/scala/kafka/server/ReplicaManager.scala:
```diff
@@ -1118,9 +1122,13 @@ class ReplicaManager(val config: KafkaConfig,
     var bytesReadable: Long = 0
     var errorReadingData = false
+
+    // The 1st topic-partition that has to be read from remote storage
+    var remoteFetchInfo: Optional[RemoteStorageFetchInfo] = Optional.empty()
```
I understand a new PR will come to overcome this, but could we provide further context (on the source code or PR) about the implications of using the first topic-partition only?

Hangleton replied:
Agreed - there are consumption patterns which diverge from the local case with this approach (that is, uneven progress across the partitions consumed from a topic [with said partitions of the same nature w.r.t. record batch size and overall size]).

It may be preferable not to diverge from the local approach and read from all the remote partitions found in the `fetchInfos`. Then, a different read pattern which provides greater performance for a specific operational environment and workload could be enforced via a configuration property.

satishd replied:
As I already called out in this PR description, that it is followed up with a PR. We will describe the config on different options with respective scenarios. The default value will be to fetch from multiple partitions as it does with local log segments.

Hangleton replied:
Got it, thanks.

jeqo replied:
Sure, thanks.

satishd committed:
Added a minor doc for a specific case

junrao commented on core/src/main/java/kafka/log/remote/RemoteLogManager.java:
```diff
@@ -600,6 +622,176 @@ public String toString() {
+        int maxBytes = Math.min(fetchMaxBytes, fetchInfo.maxBytes);
+
+        Optional<UnifiedLog> logOptional = fetchLog.apply(tp);
+        OptionalInt epoch = OptionalInt.empty();
```
Could we just use `fetchInfo.lastFetchedEpoch`?

satishd replied:
afaik, `lastFetchedEpoch` is the epoch of the last fetched record. That can be different from the fetch offset’s epoch. We should find the respective epoch for the target offset and use that to find the remote log segment metadata.

junrao commented on core/src/main/java/kafka/log/remote/RemoteLogManager.java:
```diff
@@ -600,6 +622,176 @@ public String toString() {
+                    + epochStr + " and partition " + tp + " which does not exist in remote tier.");
+        }
+
+        int startPos = lookupPositionForOffset(rlsMetadata.get(), offset);
```
We are calling `rlsMetadata.get()` multiple times. We could get it once and reuse the result?

junrao commented on core/src/main/java/kafka/log/remote/RemoteLogManager.java:
```diff
@@ -600,6 +622,176 @@ public String toString() {
+            remainingBytes -= firstBatch.sizeInBytes();
+
+            if (remainingBytes > 0) {
+                // input stream is read till (startPos - 1) while getting the batch of records earlier.
```
Hmm, not sure that I understand the comment. What is startPos?

junrao commented on core/src/main/java/kafka/log/remote/RemoteLogManager.java:
```diff
@@ -600,6 +622,176 @@ public String toString() {
+            }
+            buffer.flip();
+
+            FetchDataInfo fetchDataInfo = new FetchDataInfo(new LogOffsetMetadata(offset), MemoryRecords.readableRecords(buffer));
```
Hmm, `LogOffsetMetadata.position` is important for calculating `upperBoundOffset` in `addAbortedTransactions()`. So, we need to set it instead of relying on the default, right?

satishd replied:
Good catch, addressed it in the latest commit.

junrao commented on core/src/main/java/kafka/log/remote/RemoteLogManager.java:
```diff
@@ -600,6 +622,176 @@ public String toString() {
+                                    RemoteLogSegmentMetadata segmentMetadata,
+                                    Consumer<List<AbortedTxn>> accumulator) throws RemoteStorageException {
+        TopicPartition topicPartition = segmentMetadata.topicIdPartition().topicPartition();
+        Iterator<LogSegment> localLogSegments = fetchLog.apply(topicPartition)
```
We are calling `fetchLog.apply` multiple time when serving a single fetch request. Could we call it once and reuse the result?

satishd committed:
Fixed tests with the right APIs

junrao commented on core/src/main/java/kafka/log/remote/RemoteLogManager.java:
```diff
@@ -600,6 +622,176 @@ public String toString() {
+                }
+
+                if (searchInLocalLog) {
+                    txnIndexOpt = (localLogSegments.hasNext()) ? Optional.of(localLogSegments.next().txnIndex()) : Optional.empty();
```
localLogSegments could overlap with the remote segment. So, we could introduce duplicate aborted txn records here.

satishd replied:
Right, it can have duplicates. But consumer already handles the duplicate aborted transactions. Updated the code to remove duplicates incase any consumer implementation can not handle duplicate aborted transactions.

junrao replied:
Where is the logic to remove the duplicates?

Also, remote segments could overlap too. Is there  any logic to remove the duplicates from remote segments too?

satishd replied:
`accumulator` collects these transactions and adds them to a set that removes the duplicates. You can find the accumulator [here](https://github.com/apache/kafka/pull/13535/files#diff-380e4d8859ea9148f21794c09039425c82d9012a392c2dbbe1ce2ec8677a1970R727).

junrao commented on core/src/main/java/kafka/log/remote/RemoteLogReader.java:
```diff
@@ -0,0 +1,70 @@
+import java.util.function.Consumer;
+
+public class RemoteLogReader implements Callable<Void> {
+
```
extra new line

junrao commented on core/src/main/java/kafka/log/remote/RemoteLogReadResult.java:
```diff
@@ -0,0 +1,33 @@
+import java.util.Optional;
+
+public class RemoteLogReadResult {
+
```
extra new line

junrao commented on core/src/main/java/kafka/log/remote/RemoteLogReadResult.java:
```diff
@@ -0,0 +1,33 @@
+        this.fetchDataInfo = fetchDataInfo;
+        this.error = error;
+    }
+
```
extra new line

junrao commented on core/src/main/java/kafka/log/remote/RemoteLogReader.java:
```diff
@@ -0,0 +1,70 @@
+            result = new RemoteLogReadResult(Optional.empty(), Optional.of(e));
+        }
+
+        logger.debug("Finished reading records from remote storage for topic partition {}", fetchInfo.topicPartition);
```
Should we log fetchInfo too?

Hangleton replied:
Should we report of offset out of range when applicable?

junrao commented on core/src/main/scala/kafka/server/DelayedRemoteFetch.scala:
```diff
@@ -0,0 +1,121 @@
+            debug(s"Broker is the leader of partition $topicPartition, but the requested epoch " +
+              s"$fetchLeaderEpoch is fenced by the latest leader epoch, satisfy $fetchParams immediately")
+            return forceComplete()
+          case _: NotLeaderOrFollowerException =>  // Case a
```
Should we add follower in the comment and the log below for case a?

junrao commented on core/src/main/scala/kafka/server/DelayedRemoteFetch.scala:
```diff
@@ -0,0 +1,121 @@
+          case _: UnknownTopicOrPartitionException => // Case b
+            debug(s"Broker no longer knows of partition $topicPartition, satisfy $fetchParams immediately")
+            return forceComplete()
+          case _: FencedLeaderEpochException => // Case e
```
Hmm, `replicaManager.getPartitionOrException` doesn't seem to check the epoch?

junrao commented on core/src/main/java/kafka/log/remote/RemoteLogReader.java:
```diff
@@ -0,0 +1,70 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package kafka.log.remote;
```
Could RemoteLogReader, RemoteLogReadResult and RemoteStorageThreadPool all be in the storage module?

satishd replied:
`RemoteLogReader` can not be moved to storage module as it currently depends on `RemoteLogManager`. I will move along with `RemoteLogManager` later. 
`RemoteLogReadResult` and `RemoteStorageThreadPool` are moved to storage module.

junrao commented on core/src/main/java/kafka/log/remote/RemoteStorageThreadPool.java:
```diff
@@ -0,0 +1,76 @@
+public class RemoteStorageThreadPool extends ThreadPoolExecutor {
+    private final Logger logger;
+
+    public RemoteStorageThreadPool(String name,
```
unused param

junrao commented on core/src/main/java/kafka/log/remote/RemoteStorageThreadPool.java:
```diff
@@ -0,0 +1,76 @@
+                                   String threadNamePrefix,
+                                   int numThreads,
+                                   int maxPendingTasks,
+                                   Time time) {
```
unused param

junrao reviewed (commented):
@satishd : Thanks for the PR. A few comments below.

jeqo commented on core/src/main/java/kafka/log/remote/RemoteLogManager.java:
```diff
@@ -600,6 +622,176 @@ public String toString() {
+        InputStream remoteSegInputStream = null;
+        try {
+            // Search forward for the position of the last offset that is greater than or equal to the target offset
+            remoteSegInputStream = remoteLogStorageManager.fetchLogSegment(rlsMetadata.get(), startPos);
```
Would be possible sending the endOffset as well? Without it, input stream will potentially contain the whole log and not be consumed til the end.
In the case of S3, when inputstream is not consumed til the end HTTP connection is aborted.

satishd replied:
We will look into it in a followup PR.

jeqo replied:
Sure, thanks.

Hangleton commented on core/src/main/java/kafka/log/remote/RemoteLogManager.java:
```diff
@@ -600,6 +622,176 @@ public String toString() {
+                return new FetchDataInfo(new LogOffsetMetadata(offset), MemoryRecords.EMPTY, false,
+                        includeAbortedTxns ? Optional.of(Collections.emptyList()) : Optional.empty());
+
+            int updatedFetchSize =
```
Is there a risk here to overflow the `buffer` if it is allocated up to `maxBytes` and the first batch is greater than that?

This could happen if the partition read here is not the first one from the Fetch request in which case `minOneMessage` is false [*]. In the local counterpart, an empty set of records is returned ([source](https://github.com/apache/kafka/blob/trunk/core/src/main/scala/kafka/server/ReplicaManager.scala#L1247-L1250), where `firstEntryIncomplete` describes the case where (using the equivalent notation used here) `minOneMessage` was false and `firstBatch.sizeInBytes() > maxBytes`).

[*] Even if only one remote partition is served per Fetch request, there can still be another partition from that request served from its local replica log hence a possibility for not at least one message requested to be read here.

satishd replied:
Good point. There is no risk here but it is good to be consistent with the local read pattern to return empty records for that case.

satishd replied:
There is no risk here but it is good to be consistent with the local read pattern to return empty records for that case. Updated with the latest commit.

Hangleton commented on core/src/main/java/kafka/log/remote/RemoteLogManager.java:
```diff
@@ -600,6 +622,176 @@ public String toString() {
+                if (!searchInLocalLog) {
+                    nextSegmentMetadataOpt = findNextSegmentMetadata(nextSegmentMetadataOpt.get());
+
+                    txnIndexOpt = nextSegmentMetadataOpt.map(x -> indexCache.getIndexEntry(x).txnIndex());
```
There may be an availability risk here. The remote index cache allows up to 1,024 entries by default and it is possible to keep a transaction index entry permanently rotated in and out of the cache so that retries of Fetch requests for a partition at a given (constant) fetch offset continuously trigger a download of the index. This, plus the fact there is a global lock for read and write access on the remote index cache, can lead to lock contention on the remote log reader thread pool. In the worst case, it is possible for a partition to be prevented from making progress on the fetch path.

satishd replied:
1024 is the default value. Of course, this will be exposed as a config so that users can configure based on their workload patterns.

Hangleton replied:
Thanks, it makes sense to have it configurable. For the use case reported above, I think it's probably best to come up with a specific scenario and a test to reproduce. I can give it a shot in a couple of weeks, which will be after the PR is merged.

Hangleton commented on core/src/main/scala/kafka/server/ReplicaManager.scala:
```diff
@@ -1273,17 +1328,45 @@ class ReplicaManager(val config: KafkaConfig,
-            lastStableOffset = None,
-            exception = Some(e))
+          createLogReadResult(e)
+        case e: OffsetOutOfRangeException =>
```
I assume eventually we won't use exception-based branch control and not rely on this exception to be re-directed to the remote read code path?

satishd replied:
Yes.

satishd committed:
Addressed reveiw comments

satishd committed:
Moved RemoteLogReadResult and RemoteStorageThreadPool to storage module

satishd committed:
Addressed review comments
 - Added position and segment base offset to LogOffsetMetadata in FetchDataInfo

satishd reviewed (commented):
Thanks @junrao for your review. Addressed your comments inline and/or with the latest commits.

satishd committed:
Addressed review comments
 - Handling duplicate aborted transactions and minor cleanup.

satishd committed:
Addressed review comments
 - Sending empty records like it does with local logs when when there is no minimum-one-message constraint and for FetchRequest version 3 and above and the first batch size is more than maximum bytes that can be sent.

showuon reviewed (commented):
LGTM! @satishd , will there be other tests added as in the PR description said, or there will be follow-up PRs to add them?

divijvaidya commented on core/src/main/java/kafka/log/remote/RemoteLogManager.java:
```diff
@@ -600,25 +622,208 @@ public String toString() {
+        try {
+            // Search forward for the position of the last offset that is greater than or equal to the target offset
+            remoteSegInputStream = remoteLogStorageManager.fetchLogSegment(remoteLogSegmentMetadata, startPos);
+            RemoteLogInputStream remoteLogInputStream = new RemoteLogInputStream(remoteSegInputStream);
```
Perf comment: not blocking for this PR

Note that `RemoteLogInputStream` performs a heap allocation here for each batch. This allocation will increase GC activity with the size of segment. Perhaps, we can use BufferSupplier (we allocate one pool per request handler thread, see `requestLocal.bufferSupplier` in KafkaAPI.scala) here in future.

satishd replied:
Sure, we can definitely look into that.  We did not see much of GC issues remote read throughputs ~750 MBps on a broker but there are plans to improve by exploring buffer pool mechanisms(variations of pool used in producers).

divijvaidya replied:
Thanks, I added a JIRA https://issues.apache.org/jira/browse/KAFKA-14954 to ensure that we are tracking it and don't forget about it.

I agree that this is not a major blocker from GC perspective as JVMs are getting more intelligent about allocation/deallocation of byte[] but this adds up in situations when we have large number of allocation. Another PR where we removed data copy and allocation resulted in 2% CPU improvement (measured from flamegraph) with JDK17. JDK 8 will probably have a larger impact.

divijvaidya commented on core/src/main/java/kafka/log/remote/RemoteLogManager.java:
```diff
@@ -600,25 +622,208 @@ public String toString() {
+            // and for FetchRequest version 3 and above and the first batch size is more than maximum bytes that can be sent.
+            if (!remoteStorageFetchInfo.minOneMessage &&
+                    !remoteStorageFetchInfo.hardMaxBytesLimit &&
+                    firstBatch.sizeInBytes() > maxBytes) {
```
This and the above `firstBatch == null` condition can be combined since they produce the same result.

satishd replied:
After the recent change of extracting `firstBatch.sizeInBytes()`, the existing code looks simpler.

divijvaidya commented on core/src/main/java/kafka/log/remote/RemoteLogManager.java:
```diff
@@ -600,25 +622,208 @@ public String toString() {
+
+            int updatedFetchSize =
+                    remoteStorageFetchInfo.minOneMessage && firstBatch.sizeInBytes() > maxBytes
+                            ? firstBatch.sizeInBytes() : maxBytes;
```
Perf comment: not blocking for this PR

Please calculate `firstBatch.sizeInBytes()` once and reuse.

divijvaidya commented on core/src/main/scala/kafka/server/DelayedRemoteFetch.scala:
```diff
@@ -0,0 +1,112 @@
+      false
+  }
+
+  override def onExpiration():Unit = {
```
non blocking comment

we probably need metrics on expiration of delated remote fetch here.

satishd replied:
Yes, we have in the internal branch. There are other metrics too related to tiered storage.  They will be added in a followup PR. 

<img width="1415" alt="image" src="https://user-images.githubusercontent.com/2577761/235444150-338d3472-c511-4252-bc56-1afa31c1ad2b.png">

divijvaidya replied:
Thanks, I added a JIRA https://issues.apache.org/jira/browse/KAFKA-14953 to ensure that we don't forget this and are tracking it somewhere.

divijvaidya commented on core/src/main/scala/kafka/server/DelayedRemoteFetch.scala:
```diff
@@ -0,0 +1,112 @@
+
+  override def onExpiration():Unit = {
+    // cancel the remote storage read task, if it has not been executed yet
+    remoteFetchTask.cancel(false)
```
1. Why aren't we interrupting this with a true argument?
2. We should probably check the return and if we are unable to cancel, we should log a debug statement here.

divijvaidya commented on core/src/main/scala/kafka/server/DelayedRemoteFetch.scala:
```diff
@@ -0,0 +1,112 @@
+  override def onComplete():Unit = {
+    val fetchPartitionData = localReadResults.map { case (tp, result) =>
+      if (tp.topicPartition().equals(remoteFetchInfo.topicPartition) && remoteFetchResult.isDone
+        && result.exception.isEmpty && result.info.delayedRemoteStorageFetch.isPresent) {
```
nit

s/result.Exception.isEmpty/(result.error == Errors.NONE)

Asking because it makes the code more clear since we are using result.error in the else to populate the reponse.

divijvaidya commented on core/src/main/java/kafka/log/remote/RemoteLogManager.java:
```diff
@@ -147,6 +164,11 @@ public RemoteLogManager(RemoteLogManagerConfig rlmConfig,
         indexCache = new RemoteIndexCache(1024, remoteLogStorageManager, logDir);
         delayInMs = rlmConfig.remoteLogManagerTaskIntervalMs();
         rlmScheduledThreadPool = new RLMScheduledThreadPool(rlmConfig.remoteLogManagerThreadPoolSize());
+        remoteStorageReaderThreadPool = new RemoteStorageThreadPool(
```
important

We are not shutting this down in `RemoteLogManager#close()`

divijvaidya commented on core/src/main/scala/kafka/server/ReplicaManager.scala:
```diff
@@ -1160,48 +1171,100 @@ class ReplicaManager(val config: KafkaConfig,
+            val fetchPartitionData = logReadResults.map { case (tp, result) =>
+              val r = {
+                if (tp.topicPartition().equals(remoteFetchInfo.get.topicPartition))
+                  createLogReadResult(e)
```
Is this `RejectedExecutionException` propagated to the Consumer fetch? If yes, is this a change in the existing interface with the consumer? (please correct me if I am wrong but I am not aware of consumer handling or expecting `RejectedExecutionException` today.

satishd replied:
This error is [propagated](https://github.com/apache/kafka/blob/trunk/core/src/main/scala/kafka/server/ReplicaManager.scala#L86) as unexpected error (UnknownServerException) to the consumer client and it is already handled.

divijvaidya replied:
Thank you. That answers my question.

Hangleton replied:
Should we add a log if the nature of the error is not propagated?

divijvaidya reviewed (commented):
Overall looks good to me. One major comment about correctly shutting down the delayed fetch thread pool, otherwise looks good to me.

Hangleton commented on core/src/main/java/kafka/log/remote/RemoteLogManager.java:
```diff
@@ -600,25 +622,208 @@ public String toString() {
+            // An empty record is sent instead of an incomplete batch when there is no minimum-one-message constraint
+            // and for FetchRequest version 3 and above and the first batch size is more than maximum bytes that can be sent.
+            if (!remoteStorageFetchInfo.minOneMessage &&
+                    !remoteStorageFetchInfo.hardMaxBytesLimit &&
```
Since `remoteStorageFetchInfo.hardMaxBytesLimit` <=> Fetch request version <= 2, is there a possibility that Fetch requests v <= 2 are served from TS? If not, we can probably remove that condition on `hardMaxBytesLimit`. If yes, I think there may still be a risk of overflow for v <= 2.

Hangleton replied:
Tagging this comment as not longer valid since a code change addressed it.

Hangleton commented on core/src/main/java/kafka/log/remote/RemoteLogManager.java:
```diff
@@ -600,25 +622,208 @@ public String toString() {
+                : Optional.empty();
+
+        if (!rlsMetadataOptional.isPresent()) {
+            String epochStr = (epoch.isPresent()) ? Integer.toString(epoch.getAsInt()) : "NOT AVAILABLE";
```
nit: instead of `NOT_AVAILABLE`, maybe the message could report that the log start offset is strictly greater than the fetch offset?

satishd committed:
Addressed review comments

satishd committed:
Added missing cleaningup of DelayedRemoteFetchPurgatory

divijvaidya commented on core/src/main/java/kafka/log/remote/RemoteLogManager.java:
```diff
@@ -670,6 +875,14 @@ public void close() {
                 } catch (InterruptedException e) {
                     // ignore
                 }
+                remoteStorageReaderThreadPool.shutdownNow();
```
1. Should we try to complete the pending reads first using `shutdown()` instead of interrupting them via `shutdownNow()`? This would potentially reduce abrupt behaviour if RSM doesn't handle interruption correctly.

I will suggest the following pattern (also see `shutdownAndAwaitTermination` at https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/util/concurrent/ExecutorService.html)
```
        remoteStorageReaderThreadPool.shutdown(); // stop accepting new requests to this pool

        try {
                // wait until the pool has terminated
                if(!remoteStorageReaderThreadPool.awaitTermination(2, TimeUnit.MINUTES)) {
                    LOGGER.warn("Shutdown for remoteStorageReaderThreadPool didn't complete all tasks.");
                    remoteStorageReaderThreadPool.shutdownNow();    
                }
        } catch (InterruptedException e) {
            remoteStorageReaderThreadPool.shutdownNow();
            LOGGER.warn("Shutdown for remoteStorageReaderThreadPool interrupted");
        }

        LOGGER.info("RemoteStorageReaderThreadPool stopped.");
```

2. I would also suggest to explain why aren't we re-interrupting the current thread using Thread.currentThread().interrupt().

3. Please add warn and info logs that help debug the correct order of shutdown.

divijvaidya commented on core/src/main/java/kafka/log/remote/RemoteLogManager.java:
```diff
@@ -670,6 +875,14 @@ public void close() {
+                remoteStorageReaderThreadPool.shutdownNow();
+                //waits for 2 mins to terminate the current tasks
+                try {
+                    remoteStorageReaderThreadPool.awaitTermination(2, TimeUnit.MINUTES);
```
How did we decide on 2min. here? I don't think we should block shutdown of the broker on this over here because there are other limits associated with clean vs unclean shutdown. If we do plan to block, we should tie it to overall shutdown timeout. As an example, clean shutdown is expected to be completed in 5 min. see `lifecycleManager.controlledShutdownFuture.get(5L, TimeUnit.MINUTES)` in BrokerServer.scala.

satishd replied:
It does not require that to be completed in 5 mins. `lifecycleManager.controlledShutdownFuture` is more about processing the controlled shutdown event to the controller for that broker. It will wait for 5 mins before proceeding with other sequence of actions. But that will not get affected because of the code introduced here. 
Logging subsystem handles unclean shutdown for log segments and it would have been already finished before RemoteLogManager is closed. So, they will not get affected because of this timeout. But we can have a short duration here like 10 secs, we can revisit introducing a config if it is really needed for closing the remote log subsystem.

divijvaidya reviewed (commented):
Thank you for addressing the previous comments Satish. I have some additional ones about how we are handling shutdown.

Hangleton commented on core/src/main/java/kafka/log/remote/RemoteLogReader.java:
```diff
@@ -0,0 +1,70 @@
+            result = new RemoteLogReadResult(Optional.of(fetchDataInfo), Optional.empty());
+        } catch (OffsetOutOfRangeException e) {
+            result = new RemoteLogReadResult(Optional.empty(), Optional.of(e));
+        } catch (Exception e) {
```
Not specific to this code path  but one downside of not propagating interrupt exception from the plugins is that they are logged as exceptions in Kafka server logs. It may be preferable to avoid this by handling these interrupt as first-class exceptions. What do you think?

satishd replied:
As the issue is not related to this set of changes, we can look into it later.

Hangleton commented on core/src/test/java/kafka/log/remote/RemoteLogReaderTest.java:
```diff
@@ -0,0 +1,83 @@
+    }
+
+    @Test
+    public void testRemoteLogReaderWithError() throws RemoteStorageException, IOException {
```
Should we test cases where the plugins throw an exception?

satishd replied:
We will add more tests in followup PRs.

junrao commented on core/src/main/scala/kafka/server/ReplicaManager.scala:
```diff
@@ -1160,48 +1172,100 @@ class ReplicaManager(val config: KafkaConfig,
+        delayedRemoteFetchPurgatory.tryCompleteElseWatch(remoteFetch, Seq(key))
+      } else {
+        // If there is not enough data to respond and there is no remote data, we will let the fetch request
+        // to wait for new data.
```
let the fetch request to wait => let the fetch request wait

junrao commented on core/src/main/scala/kafka/server/ReplicaManager.scala:
```diff
@@ -1273,17 +1329,46 @@ class ReplicaManager(val config: KafkaConfig,
+          createLogReadResult(e)
+        case e: OffsetOutOfRangeException =>
+          // In case of offset out of range errors, check for remote log manager for non-compacted topics
+          // to fetch from remote storage. `log` instance should not be null here as that would have caught earlier with
```
that would have caught => that would have been caught

junrao commented on core/src/main/scala/kafka/server/ReplicaManager.scala:
```diff
@@ -1273,17 +1329,46 @@ class ReplicaManager(val config: KafkaConfig,
+          // If it is from a follower then send the offset metadata only as the data is already available in remote
+          // storage.
+          if (remoteLogManager.isDefined && log != null && log.remoteLogEnabled() &&
+            // Check that the fetch offset is with in the offset range with in the remote storage layer.
```
two occurrences
with in => within

junrao commented on core/src/main/scala/kafka/server/ReplicaManager.scala:
```diff
@@ -1273,17 +1329,46 @@ class ReplicaManager(val config: KafkaConfig,
+                leaderLogStartOffset,
+                leaderLogEndOffset,
+                followerLogStartOffset,
+                time.milliseconds,
```
Could we reuse fetchTimeMs instead of calling time.milliseconds again?

junrao commented on core/src/main/scala/kafka/server/ReplicaManager.scala:
```diff
@@ -1320,6 +1405,33 @@ class ReplicaManager(val config: KafkaConfig,
     result
   }
 
+  private def createLogReadResult(highWatermark: Long,
```
Should this be in `object ReplicaManager`?

junrao commented on core/src/main/scala/kafka/server/ReplicaManager.scala:
```diff
@@ -1320,6 +1405,33 @@ class ReplicaManager(val config: KafkaConfig,
+      exception = Some(e))
+  }
+
+  def createLogReadResult(e: Throwable): LogReadResult = {
```
Should this be in `object ReplicaManager`?

junrao commented on core/src/main/java/kafka/log/remote/RemoteLogManager.java:
```diff
@@ -600,25 +622,208 @@ public String toString() {
+                        includeAbortedTxns ? Optional.of(Collections.emptyList()) : Optional.empty());
+
+            // An empty record is sent instead of an incomplete batch when there is no minimum-one-message constraint
+            // and for FetchRequest version 3 and above and the first batch size is more than maximum bytes that can be sent.
```
This line doesn't read well.  Also, should it be put above the following line?

```
            int updatedFetchSize =
                    remoteStorageFetchInfo.minOneMessage && firstBatchSize > maxBytes ? firstBatchSize : maxBytes;

```

satishd committed:
Addressed review comments

divijvaidya reviewed (approved):
This change looks good to me! (assuming tests will be merged in separate PR)

junrao commented on core/src/main/java/kafka/log/remote/RemoteLogManager.java:
```diff
@@ -600,25 +622,208 @@ public String toString() {
+
+        TopicPartition topicPartition = segmentMetadata.topicIdPartition().topicPartition();
+        long nextSegmentBaseOffset = segmentMetadata.endOffset() + 1;
+        OptionalInt epoch = OptionalInt.of(segmentMetadata.segmentLeaderEpochs().lastEntry().getKey());
```
You mentioned the following in the other comment.
`We should find the respective epoch for the target offset and use that to find the remote log segment metadata.`

However, here, epoch seems to be for the offset before nextSegmentBaseOffset, not at nextSegmentBaseOffset?

satishd replied:
It initially tries to find a segment for that epoch and the offset. If that is not found, it gets the next available epoch by using `leaderEpochFileCache.nextEpoch(epoch.getAsInt())` and tries to find a segment for that epoch. It eventually tries to find it by going through the epoch sequence like below.

```
        while (!result.isPresent() && epoch.isPresent()) {
            result = fetchRemoteLogSegmentMetadata(topicPartition, epoch.getAsInt(), nextSegmentBaseOffset);
            epoch = leaderEpochFileCache.nextEpoch(epoch.getAsInt());
        }
```

Hangleton replied:
Wouldn't it be possible to resolve the leader epoch of the `nextSegmentBaseOffset` directly from the leader epoch cache of the partition, removing the need for the loop altogether?

satishd replied:
Good point. We can use `leaderepochCache.epochForOffset()` which we introduced for all other usages and missed replacing it here.

junrao reviewed (commented):
@satishd : Thanks for the updated PR. Haven't looked at the testing code. A few more comments.

satishd committed:
Addressed review comments

junrao commented on core/src/main/java/kafka/log/remote/RemoteLogManager.java:
```diff
@@ -600,25 +623,210 @@ public String toString() {
+            // An empty record is sent instead of an incomplete batch when
+            //  - there is no minimum-one-message constraint and
+            //  - the first batch size is more than maximum bytes that can be sent.
+            //  - for FetchRequest version 3 or above and
```
remove "and" at the end

junrao commented on core/src/test/scala/integration/kafka/server/DelayedRemoteFetchTest.scala:
```diff
@@ -0,0 +1,179 @@
+import java.util.Optional
+import java.util.concurrent.CompletableFuture
+
+import scala.collection._
```
Could we write the new test in java?

satishd replied:
We have a few more tests that need to be pulled out from the existing 2.8.x repo with other changes and refactor from EasyMock to Mockito, we can convert this test into Java in that PR.

junrao commented on core/src/test/scala/integration/kafka/server/DelayedRemoteFetchTest.scala:
```diff
@@ -0,0 +1,179 @@
+  private val fetchParams = buildFollowerFetchParams(replicaId, maxWaitMs = 500)
+
+  @Test
+  def testFetchWithFencedEpoch(): Unit = {
```
Hmm, where is logic to simulate a fenced epoch in this test?

satishd replied:
This is not meant to be fenced epoch. It is covered in another test.

junrao commented on core/src/test/scala/integration/kafka/server/DelayedRemoteFetchTest.scala:
```diff
@@ -0,0 +1,179 @@
+    assertTrue(delayedRemoteFetch.isCompleted)
+    assertEquals(topicIdPartition, actualTopicPartition.get)
+    assertTrue(fetchResultOpt.isDefined)
+    assertTrue(fetchResultOpt.isDefined)
```
This line duplicates the line before.

junrao commented on core/src/test/scala/integration/kafka/server/DelayedRemoteFetchTest.scala:
```diff
@@ -0,0 +1,179 @@
+    assertTrue(delayedRemoteFetch.isCompleted)
+    assertEquals(topicIdPartition, actualTopicPartition.get)
+    assertTrue(fetchResultOpt.isDefined)
+    assertTrue(fetchResultOpt.isDefined)
```
This line duplicates the line before.

junrao commented on core/src/test/java/kafka/log/remote/RemoteLogReaderTest.java:
```diff
@@ -0,0 +1,83 @@
+    LogOffsetMetadata logOffsetMetadata = new LogOffsetMetadata(100);
+    Records records = mock(Records.class);
+
+
```
extra new line.

satishd committed:
Addressed reveiw comments

junrao commented on core/src/test/scala/unit/kafka/coordinator/AbstractCoordinatorConcurrencyTest.scala:
```diff
@@ -156,8 +156,7 @@ object AbstractCoordinatorConcurrencyTest {
 
-  class TestReplicaManager extends ReplicaManager(
-    null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, None, null) {
+  class TestReplicaManager extends ReplicaManager(null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, None, null) {
```
Do we need to change this since it makes the line quite long?

junrao commented on core/src/main/scala/kafka/server/ReplicaManager.scala:
```diff
@@ -1153,14 +1193,15 @@ class ReplicaManager(val config: KafkaConfig,
-    //                        6) has a preferred read replica
-    if (params.maxWaitMs <= 0 || fetchInfos.isEmpty || bytesReadable >= params.minBytes || errorReadingData ||
-      hasDivergingEpoch || hasPreferredReadReplica) {
+    //                        3) fetch request does not require any data
```
`2)` should be fetch request does not want to wait. Also, the comment is written as if `1)` is also disjunctive, but it is not.

junrao reviewed (commented):
@satishd : Thanks for the updated PR. Made a pass of all files. A few more comments.

satishd committed:
Addressed reveiw comments
 - Updated the comment to make it more clear on when a fetch response needs to be sent immediately

satishd committed:
Addressed minor comments

Hangleton commented on core/src/main/java/kafka/log/remote/RemoteLogManager.java:
```diff
@@ -600,25 +623,210 @@ public String toString() {
+        return indexCache.lookupOffset(remoteLogSegmentMetadata, offset);
+    }
+
+    private FetchDataInfo addAbortedTransactions(long startOffset,
```
Do we have a unit test for this method and/or the associated functionality taken in isolation?

satishd replied:
We have the below test scenarios in 2.8.x in Scala. We plan to raise these in a followup PR as it requires a rewrite in Java and refactor from EasyMock to Mockito.

testAddAbortedTransactions
testCollectAbortedTransactionsIteratesNextRemoteSegment
testCollectAbortedTransactionsIteratesNextLocalSegment

Hangleton commented on core/src/main/java/kafka/log/remote/RemoteLogManager.java:
```diff
@@ -600,25 +623,210 @@ public String toString() {
+        long offset = fetchInfo.fetchOffset;
+        int maxBytes = Math.min(fetchMaxBytes, fetchInfo.maxBytes);
+
+        Optional<UnifiedLog> logOptional = fetchLog.apply(tp);
```
Would it be worth adding a `DEBUG` log here to help with debugging?

Hangleton commented on core/src/main/java/kafka/log/remote/RemoteLogManager.java:
```diff
@@ -600,25 +623,210 @@ public String toString() {
+            TxnIndexSearchResult searchResult = txnIndexOpt.get().collectAbortedTxns(startOffset, upperBoundOffset);
+            accumulator.accept(searchResult.abortedTransactions);
+            if (!searchResult.isComplete) {
+                if (!searchInLocalLog) {
```
Maybe we could separate the local and remote accumulation from within the same loop and segregate the branches for local/remote from the core logic?

Hangleton commented on core/src/main/java/kafka/log/remote/RemoteLogManager.java:
```diff
@@ -578,9 +595,15 @@ public void run() {
+                Optional<UnifiedLog> unifiedLogOptional = fetchLog.apply(topicIdPartition.topicPartition());
+
+                if (!unifiedLogOptional.isPresent()) {
+                    return;
```
Should we log this case at `DEBUG` level?

Hangleton commented on core/src/main/scala/kafka/server/ReplicaManager.scala:
```diff
@@ -1288,17 +1373,46 @@ class ReplicaManager(val config: KafkaConfig,
+          // storage.
+          if (remoteLogManager.isDefined && log != null && log.remoteLogEnabled() &&
+            // Check that the fetch offset is within the offset range within the remote storage layer.
+            log.logStartOffset <= offset && offset < log.localLogStartOffset()) {
```
The log start offset and local log start offset could be updated asynchronously and non-atomically. Could that be a problem?

satishd replied:
That should work fine because it will eventually throw offset out-of-range error if the target offset does not exist.

Hangleton commented on core/src/main/scala/kafka/server/ReplicaManager.scala:
```diff
@@ -1288,17 +1373,46 @@ class ReplicaManager(val config: KafkaConfig,
+            log.logStartOffset <= offset && offset < log.localLogStartOffset()) {
+            // For follower fetch requests, throw an error saying that this offset is moved to tiered storage.
+            val highWatermark = log.highWatermark
+            val leaderLogStartOffset = log.logStartOffset
```
Ditto.

satishd replied:
It is fine as the offset can always be updated, we will send whatever is the value available at that time.

satishd committed:
Addressed reveiw comments.
 - Use the existing `epochForOffset` API to find the epoch instead of iterating through segments.

jeqo commented on core/src/main/java/kafka/log/remote/RemoteLogManager.java:
```diff
@@ -600,25 +623,204 @@ public String toString() {
+
+        boolean searchInLocalLog = false;
+        Optional<RemoteLogSegmentMetadata> nextSegmentMetadataOpt = Optional.of(segmentMetadata);
+        Optional<TransactionIndex> txnIndexOpt = nextSegmentMetadataOpt.map(metadata -> indexCache.getIndexEntry(metadata).txnIndex());
```
Related to this [question](https://issues.apache.org/jira/browse/KAFKA-7739?focusedCommentId=17720955&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-17720955): 
Does RLM expects to receive a null when this index is not available on RSM, or to throw `RemoteResourceNotFoundException`?

satishd replied:
This is answered in [JIRA](https://issues.apache.org/jira/browse/KAFKA-7739?focusedCommentId=17722119&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-17722119).

satishd committed:
Addressed reveiw comments
Separated logic to collect aborted transactions from remote and local segments and other minor cleanups.

kamalcph commented on core/src/main/java/kafka/log/remote/RemoteLogManager.java:
```diff
@@ -600,25 +623,212 @@ public String toString() {
+                                                          Consumer<List<AbortedTxn>> accumulator,
+                                                          Iterator<LogSegment> localLogSegments) {
+        while (localLogSegments.hasNext()) {
+            TransactionIndex txnIndexOpt = localLogSegments.next().txnIndex();
```
txnIndexOpt -> txnIndex

satishd committed:
Addressed review comments

Hangleton commented on core/src/test/java/kafka/log/remote/RemoteLogReaderTest.java:
```diff
@@ -0,0 +1,82 @@
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+
+public class RemoteLogReaderTest {
```
Should we add a test for this use case:
```
minOneMessage = false
hardMaxBytesLimit = false
firstBatchSize > maxBytes?
```
or a combination thereof?

Maybe, please let me know if I can help with a few unit tests. Thanks!

satishd replied:
Sure, you can add more unit tests. We have a few in Scala, which will be converted to Java and raised in a followup PR later.

Hangleton reviewed (commented):
Thanks Satish (@satishd) for the continued effort. Maybe, please let me know if I can help with a few unit tests. Thanks!

Hangleton reviewed (commented):
Thanks Satish (@satishd) for the continued effort on the PR. Maybe, please let me know if I can help with a few unit tests. Thanks!

satishd committed:
Minor reformat cleanup

junrao commented:
@satishd : Thanks for the updated PR. Are the test failures related to this PR especially the following ones related to remote store?

```
[Build / JDK 11 and Scala 2.13 / kafka.server.ListOffsetsRequestWithRemoteStoreTest.testResponseIncludesLeaderEpoch()](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-13535/30/testReport/junit/kafka.server/ListOffsetsRequestWithRemoteStoreTest/Build___JDK_11_and_Scala_2_13___testResponseIncludesLeaderEpoch___2/)
[Build / JDK 8 and Scala 2.12 / kafka.server.ListOffsetsRequestWithRemoteStoreTest.testResponseIncludesLeaderEpoch()](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-13535/30/testReport/junit/kafka.server/ListOffsetsRequestWithRemoteStoreTest/Build___JDK_8_and_Scala_2_12___testResponseIncludesLeaderEpoch__/)
[Build / JDK 8 and Scala 2.12 / kafka.server.ListOffsetsRequestWithRemoteStoreTest.testResponseIncludesLeaderEpoch()](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-13535/30/testReport/junit/kafka.server/ListOffsetsRequestWithRemoteStoreTest/Build___JDK_8_and_Scala_2_12___testResponseIncludesLeaderEpoch___2/)
```

satishd commented:
@junrao Looks like those test failures are related to this change. They do not fail on the laptop or other hosts. We are looking into them.

kamalcph committed:
Capture the error code in testResponseIncludesLeaderEpoch to investigate the test failure in Jenkins.

kamalcph commented:
> @satishd : Thanks for the updated PR. Are the test failures related to this PR especially the following ones related to remote store?
> 
> ```
> [Build / JDK 11 and Scala 2.13 / kafka.server.ListOffsetsRequestWithRemoteStoreTest.testResponseIncludesLeaderEpoch()](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-13535/30/testReport/junit/kafka.server/ListOffsetsRequestWithRemoteStoreTest/Build___JDK_11_and_Scala_2_13___testResponseIncludesLeaderEpoch___2/)
> [Build / JDK 8 and Scala 2.12 / kafka.server.ListOffsetsRequestWithRemoteStoreTest.testResponseIncludesLeaderEpoch()](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-13535/30/testReport/junit/kafka.server/ListOffsetsRequestWithRemoteStoreTest/Build___JDK_8_and_Scala_2_12___testResponseIncludesLeaderEpoch__/)
> [Build / JDK 8 and Scala 2.12 / kafka.server.ListOffsetsRequestWithRemoteStoreTest.testResponseIncludesLeaderEpoch()](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-13535/30/testReport/junit/kafka.server/ListOffsetsRequestWithRemoteStoreTest/Build___JDK_8_and_Scala_2_12___testResponseIncludesLeaderEpoch___2/)
> ```

It is not because of the changes in the PR. #10389 attempted to stabilize this test but it can still fail if the machine is slow.

junrao commented on core/src/main/scala/kafka/server/ReplicaManager.scala:
```diff
@@ -1175,49 +1216,101 @@ class ReplicaManager(val config: KafkaConfig,
-      delayedFetchPurgatory.tryCompleteElseWatch(delayedFetch, delayedFetchKeys)
+
+      if (remoteFetchInfo.isPresent) {
+        val key = new TopicPartitionOperationKey(remoteFetchInfo.get.topicPartition.topic(), remoteFetchInfo.get.topicPartition.partition())
```
`fetchMessages` is getting a bit large now. Would it be better to put the logic in the `if` part to a separate method?

junrao commented on core/src/main/scala/kafka/server/ReplicaManager.scala:
```diff
@@ -1288,17 +1373,46 @@ class ReplicaManager(val config: KafkaConfig,
-            exception = Some(e))
+          createLogReadResult(e)
+        case e: OffsetOutOfRangeException =>
+          // In case of offset out of range errors, check for remote log manager for non-compacted topics
```
The logic for handling OffsetOutOfRangeException is getting a bit large now. Would it be better to put that into a separate method?

junrao reviewed (approved):
@kamalcph : Thanks for the investigation. The PR LGTM. Just a couple of minor comments.

Also, should we reopen https://issues.apache.org/jira/browse/KAFKA-12384 since it's still flaky?

satishd committed:
Addressed review comments

satishd commented:
Thanks @junrao for the updated review. Addressed your latest minor review comments.

junrao commented on core/src/main/scala/kafka/server/ReplicaManager.scala:
```diff
@@ -1175,49 +1260,78 @@ class ReplicaManager(val config: KafkaConfig,
+
+      if (remoteFetchInfo.isPresent) {
+        val maybeLogReadResultWithError = processRemoteFetch(remoteFetchInfo.get(), params, responseCallback, logReadResults, fetchPartitionStatus)
+        if(maybeLogReadResultWithError.isDefined) {
```
space after `if `

junrao commented on core/src/main/scala/kafka/server/ReplicaManager.scala:
```diff
@@ -1175,49 +1260,78 @@ class ReplicaManager(val config: KafkaConfig,
+          // that we couldn't read from remote storage
+          val partitionToFetchPartitionData = buildPartitionToFetchPartitionData(logReadResults, remoteFetchInfo.get().topicPartition, maybeLogReadResultWithError.get)
+          responseCallback(partitionToFetchPartitionData)
+          return
```
It seems that `return` is unnecessary here?

junrao reviewed (commented):
@satishd : Thanks for the updated PR. A couple of minor comments.

satishd committed:
Addressed review comments

junrao reviewed (approved):
@satishd : Thanks for the updated PR. LGTM

Hangleton reviewed (commented):
Thanks for the PR Satish!

dajac commented:
@satishd I see many failed tests [here](https://ci-builds.apache.org/blue/organizations/jenkins/Kafka%2Fkafka-pr/detail/PR-13535/33/tests). Are they related to changes made in this PR?

satishd commented:
@dajac They do not seem to be related to this PR. Please take a look at the [comment](https://github.com/apache/kafka/pull/13535#issuecomment-1549993361).

dajac commented:
@satishd `testResponseIncludesLeaderEpoch` fails locally. Does it pass for you? It does not seem to be related to slow CI.

satishd commented:
@dajac It is passed locally on my laptop.

```
Gradle Test Run :core:test > Gradle Test Executor 65 > ListOffsetsRequestTest > testResponseDefaultOffsetAndLeaderEpochForAllVersions() PASSED

Gradle Test Run :core:test > Gradle Test Executor 65 > ListOffsetsRequestTest > testListOffsetsMaxTimeStampOldestVersion() PASSED

Gradle Test Run :core:test > Gradle Test Executor 65 > ListOffsetsRequestTest > testListOffsetsErrorCodes() PASSED

Gradle Test Run :core:test > Gradle Test Executor 65 > ListOffsetsRequestTest > testCurrentEpochValidation() PASSED

Gradle Test Run :core:test > Gradle Test Executor 65 > ListOffsetsRequestTest > testResponseIncludesLeaderEpoch() PASSED

BUILD SUCCESSFUL in 1m 8s
55 actionable tasks: 6 executed, 49 up-to-date
➜  kafka git:(apache-trunk) date
Tue May 23 18:00:01 IST 2023
➜  kafka git:(apache-trunk) git rev-parse --verify HEAD
15f8705246e094f7825b76a38d9f12f95d626ee5
➜  kafka git:(apache-trunk)
```

```
> Task :core:test

Gradle Test Run :core:test > Gradle Test Executor 71 > ListOffsetsRequestWithRemoteStoreTest > testResponseDefaultOffsetAndLeaderEpochForAllVersions() PASSED

Gradle Test Run :core:test > Gradle Test Executor 71 > ListOffsetsRequestWithRemoteStoreTest > testListOffsetsMaxTimeStampOldestVersion() PASSED

Gradle Test Run :core:test > Gradle Test Executor 71 > ListOffsetsRequestWithRemoteStoreTest > testListOffsetsErrorCodes() PASSED

Gradle Test Run :core:test > Gradle Test Executor 71 > ListOffsetsRequestWithRemoteStoreTest > testCurrentEpochValidation() PASSED

Gradle Test Run :core:test > Gradle Test Executor 71 > ListOffsetsRequestWithRemoteStoreTest > testResponseIncludesLeaderEpoch() PASSED

BUILD SUCCESSFUL in 1m 9s
55 actionable tasks: 6 executed, 49 up-to-date
➜  kafka git:(apache-trunk) date
Tue May 23 18:05:20 IST 2023
➜  kafka git:(apache-trunk) git rev-parse --verify HEAD
15f8705246e094f7825b76a38d9f12f95d626ee5
➜  kafka git:(apache-trunk)
```

dajac commented:
@satishd Weird... It fails all the time on my laptop.

```
Gradle Test Run :core:test > Gradle Test Executor 9 > ListOffsetsRequestTest > testResponseIncludesLeaderEpoch() FAILED
    org.opentest4j.AssertionFailedError: expected: <(10,1,0)> but was: <(-1,-1,78)>
        at app//org.junit.jupiter.api.AssertionFailureBuilder.build(AssertionFailureBuilder.java:151)
        at app//org.junit.jupiter.api.AssertionFailureBuilder.buildAndThrow(AssertionFailureBuilder.java:132)
        at app//org.junit.jupiter.api.AssertEquals.failNotEqual(AssertEquals.java:197)
        at app//org.junit.jupiter.api.AssertEquals.assertEquals(AssertEquals.java:182)
        at app//org.junit.jupiter.api.AssertEquals.assertEquals(AssertEquals.java:177)
        at app//org.junit.jupiter.api.Assertions.assertEquals(Assertions.java:1142)
        at app//kafka.server.ListOffsetsRequestTest.testResponseIncludesLeaderEpoch(ListOffsetsRequestTest.scala:210)
```

```
% git rev-parse --verify HEAD
15f8705246e094f7825b76a38d9f12f95d626ee5
```

dajac commented:
@satishd I raised a PR to fix this: https://github.com/apache/kafka/pull/13747. Could you take a look?

dajac commented on core/src/main/scala/kafka/server/ReplicaManager.scala:
```diff
@@ -1121,21 +1154,69 @@ class ReplicaManager(val config: KafkaConfig,
-    quota: ReplicaQuota,
-    responseCallback: Seq[(TopicIdPartition, FetchPartitionData)] => Unit
-  ): Unit = {
+  def fetchMessages(params: FetchParams,
```
A small comment. We should avoid completely changing the code style without reasons. The format of the method was not a mistake. It is the format that we mainly used in this class nowadays.

dajac commented on core/src/main/scala/kafka/server/ReplicaManager.scala:
```diff
@@ -1335,6 +1433,50 @@ class ReplicaManager(val config: KafkaConfig,
     result
   }
 
+  private def handleOffsetOutOfRangeError(tp: TopicIdPartition, params: FetchParams, fetchInfo: PartitionData,
+                                          adjustedMaxBytes: Int, minOneMessage:
+                                          Boolean, log: UnifiedLog, fetchTimeMs: Long,
+                                          exception: OffsetOutOfRangeException): LogReadResult = {
```
We usually don't format method like this. Could we put one argument per line?

dajac commented on core/src/main/scala/kafka/server/ReplicaManager.scala:
```diff
@@ -1335,6 +1433,50 @@ class ReplicaManager(val config: KafkaConfig,
+      } else {
+        // For consume fetch requests, create a dummy FetchDataInfo with the remote storage fetch information.
+        // For the first topic-partition that needs remote data, we will use this information to read the data in another thread.
+        val fetchDataInfo =
+        new FetchDataInfo(new LogOffsetMetadata(offset), MemoryRecords.EMPTY, false, Optional.empty(),
```
nit: `new FetchDataInfo` should be on previous line or indented.

chia7712 commented on storage/src/main/java/org/apache/kafka/storage/internals/log/RemoteStorageThreadPool.java:
```diff
@@ -0,0 +1,73 @@
+                new RemoteStorageThreadFactory(threadNamePrefix));
+        logger = new LogContext() {
+            @Override
+            public String logPrefix() {
```
@satishd pardon me, why we don't use the constructor to set prefix? the method `logger` does not use the value from override method, and hence the prefix is never propagated to logger.

I have filed https://issues.apache.org/jira/browse/KAFKA-17168 for it, and please feel free to correct me. Maybe I misunderstand this override method

satishd replied:
@chia7712 I think it was added as part of `logPrefix()` method implementation but not as part of the constructor because the logger should have the invoking thread name as the prefix but not the thread name that created this thread pool.

kamalcph replied:
why do we have to prefix the thread name in the logger? The user can include (%t) in conversion pattern to print the thread name in the logger.

chia7712 replied:
> because the logger should have the invoking thread name as the prefix but not the thread name that created this thread pool.

That is a good reason ( if they don't want to use `ConversionPattern` in log4j properties). However, I'm not sure whether overriding `logPrefix` can address that since we create `Logger` by the inner variable rather the override method.

It seems the follow-up should use `LogContext()`. As @kamalcph mentioned, using `ConversionPattern` should be good enough.

satishd replied:
Good points @chia7712  and @kamalcph. I agree users can always set a conversion pattern that includes thread-name, but I do not remember any other reason why this prefix was added. :) 

I suggest completely removing logPrefix from the new code instead of using the name of the thread that created threadpool which is not useful anyway.

kamalcph replied:
Opened #16657 to address this comment.
